File: flutter-boilerplate/lib/lib/currency.dart
import 'package:intl/intl.dart';

import '../constants/currency.dart';

const _localeMap = {
  'USD': 'en_US',
  'EUR': 'de_DE',
  'TRY': 'tr_TR',
};

/// Narrows a currency string from an API response to a known currency code,
/// falling back to USD for anything this app doesn't have a locale for.
String toCurrencyCode(String currency) {
  final upper = currency.toUpperCase();
  return CurrencyConstants.supportedCurrencies.contains(upper) ? upper : 'USD';
}

String formatCurrency(int cents, String currency) {
  final locale = _localeMap[currency] ?? _localeMap['USD']!;
  // simpleCurrency (unlike the plain .currency constructor) infers the real
  // "$"/"€"/"₺" glyph from locale+currency — .currency falls back to using
  // the ISO code itself as the symbol ("USD9.99") unless one is given.
  final format = NumberFormat.simpleCurrency(locale: locale, name: currency);
  return format.format(cents / 100);
}

/// A recurring plan price — same formatting as [formatCurrency], plus the
/// "/mo" cadence and a "Free" special case. Not for one-time amounts like
/// invoice line items; use [formatCurrency] for those.
String formatPrice(int cents, String currency) {
  if (cents == 0) return 'Free';
  return '${formatCurrency(cents, currency)}/mo';
}

File: tanstack-boilerplate/src/lib/currency.ts
import { CURRENCIES, type CurrencyCode } from "@/constants/currency";

const LOCALE_MAP: Record<CurrencyCode, string> = {
  USD: "en-US",
  EUR: "de-DE",
  TRY: "tr-TR",
};

/** Narrows a currency string from an API response to a known CurrencyCode,
 * falling back to USD for anything this app doesn't have a locale for. */
export function toCurrencyCode(currency: string): CurrencyCode {
  const upper = currency.toUpperCase();
  return (CURRENCIES as readonly string[]).includes(upper)
    ? (upper as CurrencyCode)
    : "USD";
}

export function formatCurrency(cents: number, currency: CurrencyCode): string {
  const locale = LOCALE_MAP[currency];
  return new Intl.NumberFormat(locale, {
    style: "currency",
    currency,
  }).format(cents / 100);
}

/** A recurring plan price — same formatting as formatCurrency, plus the
 * "/mo" cadence and a "Free" special case. Not for one-time amounts like
 * invoice line items; use formatCurrency for those. */
export function formatPrice(
  cents: number,
  currency: CurrencyCode,
  freeLabel: string,
): string {
  if (cents === 0) return freeLabel;
  return `${formatCurrency(cents, currency)}/mo`;
}
